{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e98d53aa",
   "metadata": {},
   "source": [
    "# QIIME2 Filtering Options for Feature Table and Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ac8d497",
   "metadata": {},
   "source": [
    "This notebook showcases different filtering options that could be performed after generating feature table and representative sequences either through clustering or denoising. Filtering your data further could be helpful in reducing the noise in your data. Make sure to choose the filtering options and parameters that are suitable for your objectives.\n",
    "\n",
    "Other filtering examples (filtering of singletons and chimeras) could also be found in the <font face=\"Consolas\">**Amplicon_Clustering_Pipeline.ipynb**</font> Jupyter notebook.\n",
    "\n",
    "This notebook is mainly based on <a href = 'https://github.com/LangilleLab/microbiome_helper/wiki/Amplicon-SOP-v2-(qiime2-2020.8)'>LangilleLab SOP</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aa65306",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## <font color ='blue'>How to Use This Notebook</font>\n",
    "\n",
    "1. Activate conda environment in terminal window. Make sure to change the environment name to what is applicable in your case.\n",
    ">`conda activate qiime2-2021.4`\n",
    "2. Open jupyter notebook with the command below and select the notebook.\n",
    ">`jupyter-notebook`\n",
    "3. To run the cells in this notebook, press Shift+Enter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3328412",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Tools Used\n",
    "1. <b>QIIME2 2021.4</b>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6160de32",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Starting Files \n",
    "\n",
    "1. This Jupyter notebook.\n",
    "2. Feature table and sequences which will be copied from the <font face=\"Consolas\">**amplicon_sample_data/asv_table_and_sequences**</font> using the code block below. The sample data used here are the same outputs of the <font face=\"Consolas\">**Amplicon_Denoising_Pipeline.ipynb**</font> jupyter notebook.\n",
    "3. Directories for organizing the data. To make the folders and copy the sample data, run the following code blocks:`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee4689b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create demo directories\n",
    "!mkdir data_filtering_demo_folder\n",
    "%cd data_filtering_demo_folder\n",
    "!mkdir 0-feature_table_and_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ad2f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ../amplicon_sample_data/asv_table_and_sequences/* 0-feature_table_and_sequences/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b889883",
   "metadata": {},
   "source": [
    "---\n",
    "## Table of Contents\n",
    " * [**1: Filtering Rare Features**](#1:-Filtering-Rare-Features)  \n",
    " * [**2: Filter Taxons and Unclassified Features**](#2:-Filter-Taxons-and-Unclassified-Features)  \n",
    " * [**3: Exclude Low-depth Samples**](#3:-Exclude-Low-depth-Samples)\n",
    "     * [Generate Rarefaction Curves](#Generate-Rarefaction-Curves)\n",
    "     * [Filter Based on Rarefaction Curves](#Filter-Based-on-Rarefaction-Curves)\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e95c960",
   "metadata": {},
   "source": [
    "# <font color = 'gray'>1: Filtering Rare Features</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6fdf8dc",
   "metadata": {},
   "source": [
    "The first filtering step will remove features that have low frequency. The important option here is the <font face = 'Consolas'><b>--p-min-frequency</b></font> which will specify the minimum frequency of a feature to be included. For Miseq data, a good value would be 0.1% of the mean frequency to remove ASVs that are likely generated by bleed-through between runs.\n",
    "\n",
    "To calculate this, check the visual output of the feature table (<font face = 'Consolas'>**2-table-dada2.qzv**</font>). Inside the visual output, go the \"Frequency per sample\" section, check the mean frequency, and multiply this by 0.001. Then replace the value specified in the <font face = 'Consolas'><b>--p-min-frequency</b></font> option.\n",
    "\n",
    "However, if you think you're losing too much features, you can be more lenient. For example, you can opt to remove singletons similar to the one performed in the <font face=\"Consolas\">**Amplicon_Clustering_Pipeline.ipynb**</font> notebook.\n",
    "\n",
    "<font color = 'red'>NOTE: Replace the value specified in <font face = 'Consolas'><b>--p-min-frequency</b></font> by what is applicable in your case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6d7857",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize 2-table-dada2.qzv\n",
    "import qiime2 as q2\n",
    "q2.Visualization.load('0-feature_table_and_sequences/2-table-dada2.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56430d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter the table\n",
    "!qiime feature-table filter-features \\\n",
    "    --i-table 0-feature_table_and_sequences/2-table-dada2.qza \\\n",
    "    --p-min-frequency 8 \\\n",
    "    --p-min-samples 1 \\\n",
    "    --o-filtered-table 1-filt-table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5f40800",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter the sequences as well\n",
    "!qiime feature-table filter-seqs \\\n",
    "    --i-data 0-feature_table_and_sequences/2-rep-seqs-dada2.qza \\\n",
    "    --i-table 1-filt-table.qza \\\n",
    "    --o-filtered-data 1-filt-seqs.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d093afcc",
   "metadata": {},
   "source": [
    "Now, let's create visualization files to see how many features are left."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d483c8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create .qzv file of filtered table and sequences\n",
    "!qiime feature-table summarize \\\n",
    "    --i-table 1-filt-table.qza \\\n",
    "    --o-visualization 1-filt-table.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "    --i-data 1-filt-seqs.qza \\\n",
    "    --o-visualization 1-filt-seqs.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25dbf943",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered table\n",
    "q2.Visualization.load('1-filt-table.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba63cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered sequences\n",
    "q2.Visualization.load('1-filt-seqs.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ef9b3d",
   "metadata": {},
   "source": [
    "---\n",
    "# <font color = 'gray'>2: Filter Taxons and Unclassified Features</font>\n",
    "\n",
    "The second filtering step will remove features that are classified as coming from certain taxonomic groups (in the example below, the putative metazoan and fungal sequences). Optionally, for example, you may also want to remove features that are unclassified at the phylum level (indicated in the <b><font face = 'Consolas'>--p-include</font></b> option). However, if you are working with a poorly characterized environment or searching for a novel taxon, it may be best to retain unclassified features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef92879",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter the specified taxon labels on the feature table\n",
    "!qiime taxa filter-table \\\n",
    "    --i-table 0-feature_table_and_sequences/2-table-dada2.qza \\\n",
    "    --i-taxonomy 0-feature_table_and_sequences/1-asv-taxa.qza \\\n",
    "    --p-exclude p__Metazoa,p__Fungi,p__Porifera,p__Cnidaria,p__Lophophorata,p__Platyhelminthes \\\n",
    "    --p-include p__ \\\n",
    "    --o-filtered-table 2-filt-table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74325ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter the specified taxon labels on the feature sequences\n",
    "!qiime taxa filter-seqs \\\n",
    "    --i-sequences 0-feature_table_and_sequences/2-rep-seqs-dada2.qza \\\n",
    "    --i-taxonomy 0-feature_table_and_sequences/1-asv-taxa.qza \\\n",
    "    --p-exclude p__Metazoa,p__Fungi,p__Porifera,p__Cnidaria,p__Lophophorata,p__Platyhelminthes \\\n",
    "    --p-include p__ \\\n",
    "    --o-filtered-sequences 2-filt-seqs.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d99986fd",
   "metadata": {},
   "source": [
    "To check whether the specified sequences are removed, we can use the taxa barplot. If we set the taxonomic level to 2, we would no longer see unclassified sequences (at the phylum level) and the other sequences specified in the <font face=\"Consolas\">**--p-exclude**</font> option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02cdb210",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate taxa barplot post-filtering\n",
    "!qiime taxa barplot \\\n",
    "    --i-table 2-filt-table.qza \\\n",
    "    --i-taxonomy 0-feature_table_and_sequences/1-asv-taxa.qza \\\n",
    "    --o-visualization 2-taxa-barplot.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36603844",
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize taxa barplot after filtering\n",
    "q2.Visualization.load('2-taxa-barplot.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9706a32e",
   "metadata": {},
   "source": [
    "Now, let's create visualization files of the filtered data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8a4fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create .qzv file of filtered table and sequences\n",
    "!qiime feature-table summarize \\\n",
    "    --i-table 2-filt-table.qza \\\n",
    "    --o-visualization 2-filt-table.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "    --i-data 2-filt-seqs.qza \\\n",
    "    --o-visualization 2-filt-seqs.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14c734c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered table\n",
    "q2.Visualization.load('2-filt-table.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ebb0996",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered sequences\n",
    "q2.Visualization.load('2-filt-seqs.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd303c86",
   "metadata": {},
   "source": [
    "---\n",
    "# <font color = 'gray'>3: Exclude Low-depth Samples</font>\n",
    "\n",
    "The final filtering example will exclude **samples** based on a specified minimum frequency of features (<b><font face = 'Consolas'>--p-min-frequency</font></b>). In choosing the value for this option, the guides below will be using a value that is based on the rarefaction curves. This will help us assess at what sampling depth does the diversity of our samples start to level out. Another possible value you could supply to the minimum frequency cutoff is 2000, which, based on this <a href = 'https://forum.qiime2.org/t/deep-sequencing/3586/2'>forum</a>, is enough to capture community diversity. However, each dataset is unique and it is best try different things and see what works for your own application. \n",
    "\n",
    "If you think you are losing too much samples, you may decrease the minimum cutoff value. You may also opt to skip these steps if you do not want to exclude any samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b35df13",
   "metadata": {},
   "source": [
    "First, let us visualize our feature table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf75819",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize\n",
    "q2.Visualization.load('0-feature_table_and_sequences/2-table-dada2.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15751ac",
   "metadata": {},
   "source": [
    "### Generate Rarefaction Curves"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77cc5976",
   "metadata": {},
   "source": [
    "Once viewed, look for the sample with highest depth (in the <i>Interactive Sample Detail</i>)  and use its depth as the input for the <b><font face = 'Consolas'>--p-max-depth</font></b> option to generate the rarefaction curves.\n",
    "\n",
    "<font color = 'red'>NOTE: Replace the <font face = 'Consolas'><b>--p-max-depth</b></font> based on the summary visualization file of <font face = 'Consolas'><b>2-table-dada2.qzv</b></font>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5b207fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generating the rarefaction curves and viewing the result\n",
    "!qiime diversity alpha-rarefaction \\\n",
    "    --i-table 0-feature_table_and_sequences/2-table-dada2.qza \\\n",
    "    --p-max-depth 12760 \\\n",
    "    --p-steps 50 \\\n",
    "    --p-metrics \"shannon\" \\\n",
    "    --o-visualization 3-rrfctn_curves.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1d72e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize\n",
    "q2.Visualization.load('3-rrfctn_curves.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "389e4170",
   "metadata": {},
   "source": [
    "### Filter Based on Rarefaction Curves"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dce72aa",
   "metadata": {},
   "source": [
    "Finally, inspect the generated rarefaction curves and determine the depth at which the rarefaction curves start to level out and use this as the minimum cutoff for sample filtering. In this case, 1500 seems to be a good value where all curves plateaued already. Moreover, since no sample has sampling depth below 1500, none will be filtered out.\n",
    "\n",
    "<font color = 'red'>NOTE: Replace the <font face = 'Consolas'><b>--p-min-frequency</b></font> option based on the rarefaction curves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f545d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filtering based on the a specified minimum frequency\n",
    "!qiime feature-table filter-samples \\\n",
    "    --i-table 0-feature_table_and_sequences/2-table-dada2.qza \\\n",
    "    --p-min-frequency 1500 \\\n",
    "    --o-filtered-table 3-filt-table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bca6d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter the sequences as well\n",
    "!qiime feature-table filter-seqs \\\n",
    "    --i-data 0-feature_table_and_sequences/2-rep-seqs-dada2.qza \\\n",
    "    --i-table 3-filt-table.qza \\\n",
    "    --o-filtered-data 3-filt-seqs.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d1c0660",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create .qzv file\n",
    "!qiime feature-table summarize \\\n",
    "    --i-table 3-filt-table.qza \\\n",
    "    --o-visualization 3-filt-table.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "    --i-data 3-filt-seqs.qza \\\n",
    "    --o-visualization 3-filt-seqs.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9679edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered table\n",
    "q2.Visualization.load('3-filt-table.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c1f792",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize filtered table\n",
    "q2.Visualization.load('3-filt-seqs.qzv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
